File: sdk-go/dex/options.go
// Legacy Materials in this file remain under their original licenses.
// See LEGACY_NOTICES.md.

// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications after the Legacy Cutoff are licensed under the
// Super Durable Source License 1.0.
// Legacy Materials remain under their original licenses.
// See LICENSE and LEGACY_NOTICES.md.

package dex

import "time"

// RetryPolicy overrides retry timing for a Step's WaitFor or Execute method.
// Zero fields preserve server defaults, including a four-hour total duration.
// MaximumAttempts counts the initial attempt.
// With asynchronous durability, local and regular execution share attempts and elapsed duration.
// Fallback is immediate, while subsequent regular retries continue the cumulative backoff sequence.
type RetryPolicy struct {
	// InitialInterval is the delay before the first retry.
	InitialInterval time.Duration
	// BackoffCoefficient multiplies each successive retry interval.
	BackoffCoefficient float64
	// MaximumInterval caps the delay between retries.
	MaximumInterval time.Duration
	// MaximumAttempts limits total attempts, including the initial attempt.
	MaximumAttempts int32
	// TotalDuration limits elapsed time across all attempts.
	TotalDuration time.Duration
}

// StepDurability controls when staged persistence must be durably acknowledged.
type StepDurability uint8

const (
	// StepDurabilityDefault uses the Flow or server default.
	StepDurabilityDefault StepDurability = iota
	// StepDurabilitySync waits for persistence before completing the handler request.
	StepDurabilitySync
	// StepDurabilityAsync permits persistence after accepting the handler result.
	StepDurabilityAsync
)

// WaitForFailurePolicy selects behavior after WaitFor exhausts its retry policy.
type WaitForFailurePolicy uint8

const (
	// FailFlowOnWaitForFailure ends the Flow after WaitFor exhausts retries.
	FailFlowOnWaitForFailure WaitForFailurePolicy = iota
	// ProceedOnWaitForFailure invokes Execute and exposes the failure through Context.
	ProceedOnWaitForFailure
)

// ExecuteFailure identifies a recovery movement after Execute exhausts retries.
// Create it with ProceedToOnExecuteFailure.
type ExecuteFailure struct {
	step    StepDef
	options *StepOptions
}

// ProceedToOnExecuteFailure routes an Execute failure to step with the same typed input.
// options override the recovery Step configuration for this movement; nil uses registered options.
func ProceedToOnExecuteFailure[IN any](
	step Step[IN],
	options *StepOptions,
) *ExecuteFailure {
	return &ExecuteFailure{
		step:    typedStepDef[IN]{step: step},
		options: options,
	}
}

// FlowTimeoutHandlerFailure identifies a recovery Step after a timeout handler exhausts retries.
// Create it with ProceedToOnFlowTimeoutHandlerFailure. The target must be a Step[None]
// registered by the timed-out Flow. The target reads the final handler error from Context.
type FlowTimeoutHandlerFailure struct {
	step    StepDef
	options *StepOptions
}

// ProceedToOnFlowTimeoutHandlerFailure routes a timeout handler failure to a no-input Step.
// options override the recovery Step configuration for this movement; nil uses registered options.
// The recovery Step receives None and reads the final handler error from Context.
func ProceedToOnFlowTimeoutHandlerFailure(
	step Step[None],
	options *StepOptions,
) *FlowTimeoutHandlerFailure {
	return &FlowTimeoutHandlerFailure{
		step:    typedStepDef[None]{step: step},
		options: options,
	}
}

// StepOptions configures one Step's handler execution and persistence behavior.
//
// Zero values preserve server defaults. Regular attempts default to two hours with a one-minute
// heartbeat timeout. Durability resolves from the method override, FlowConfig, then synchronous.
// Asynchronous durability first uses at most seven local-activity seconds and three attempts.
// WaitFor failures fail the Flow by default. Each lock is held only for the matching handler call.
type StepOptions struct {
	// WaitForMethodTimeout limits one WaitFor attempt; zero uses the server default.
	WaitForMethodTimeout time.Duration
	// ExecuteMethodTimeout limits one Execute attempt; zero uses the server default.
	ExecuteMethodTimeout time.Duration
	// HeartbeatTimeout detects stalled regular WaitFor and Execute activities.
	// Zero uses the one-minute default. Positive values must be whole seconds within int32 range
	// and meet the server-configured minimum, which defaults to ten seconds.
	HeartbeatTimeout time.Duration
	// WaitForRetry overrides the WaitFor retry policy; nil uses server defaults.
	WaitForRetry *RetryPolicy
	// ExecuteRetry overrides the Execute retry policy; nil uses server defaults.
	ExecuteRetry *RetryPolicy
	// WaitForFailure selects exhausted WaitFor behavior; zero fails the Flow.
	WaitForFailure WaitForFailurePolicy
	// ExecuteFailure selects a recovery Step after Execute exhausts retries.
	ExecuteFailure *ExecuteFailure
	// WaitForDurability overrides persistence durability for WaitFor writes.
	WaitForDurability StepDurability
	// ExecuteDurability overrides persistence durability for Execute writes.
	ExecuteDurability StepDurability
	// WaitForLockAttributes are acquired together for the WaitFor invocation.
	WaitForLockAttributes []AttributeLock
	// ExecuteLockAttributes are acquired together for the Execute invocation.
	ExecuteLockAttributes []AttributeLock
	// WaitForLoadAttributeMaps includes every current instance of each AttributeMap in WaitFor.
	WaitForLoadAttributeMaps []AttributeDef
	// WaitForLoadAttributeMapInstances includes exact AttributeMap instances in WaitFor.
	WaitForLoadAttributeMapInstances []AttributeMapLoad
	// WaitForLoadChannels includes pending messages from selected Channels in WaitFor.
	WaitForLoadChannels []ChannelDef
	// WaitForLoadChannelMaps includes every current ChannelMap instance in WaitFor.
	WaitForLoadChannelMaps []ChannelDef
	// WaitForLoadChannelMapInstances includes exact ChannelMap instance messages in WaitFor.
	WaitForLoadChannelMapInstances []ChannelMapLoad
	// ExecuteLoadAttributeMaps includes every current instance of each AttributeMap in Execute.
	ExecuteLoadAttributeMaps []AttributeDef
	// ExecuteLoadAttributeMapInstances includes exact AttributeMap instances in Execute.
	ExecuteLoadAttributeMapInstances []AttributeMapLoad
	// ExecuteLoadChannels includes pending messages from selected Channels in Execute.
	ExecuteLoadChannels []ChannelDef
	// ExecuteLoadChannelMaps includes every current ChannelMap instance in Execute.
	ExecuteLoadChannelMaps []ChannelDef
	// ExecuteLoadChannelMapInstances includes exact ChannelMap instance messages in Execute.
	ExecuteLoadChannelMapInstances []ChannelMapLoad
}

// FlowTimeoutHandlerOptions configures timeout-handler execution and state loading.
// Zero values preserve the server's Execute defaults. Ordinary Attributes and Channel size
// metadata are loaded automatically. AttributeMap values and pending Channel messages require
// an explicit load selection. One logical handler execution may contain multiple retry attempts.
type FlowTimeoutHandlerOptions struct {
	// MethodTimeout limits one timeout-handler attempt; zero uses the server default.
	MethodTimeout time.Duration
	// HeartbeatTimeout detects a stalled regular timeout-handler attempt.
	// Zero uses the server default. Positive values must be whole seconds within int32 range.
	HeartbeatTimeout time.Duration
	// Retry overrides the timeout-handler retry policy; nil uses server defaults.
	Retry *RetryPolicy
	// Failure selects a no-input recovery Step after the handler exhausts retries.
	Failure *FlowTimeoutHandlerFailure
	// Durability overrides persistence durability for timeout-handler writes.
	Durability StepDurability
	// LockAttributes are acquired together for the timeout-handler invocation.
	LockAttributes []AttributeLock
	// LoadAttributeMaps includes every current instance of each AttributeMap.
	LoadAttributeMaps []AttributeDef
	// LoadAttributeMapInstances includes exact AttributeMap instances.
	LoadAttributeMapInstances []AttributeMapLoad
	// LoadChannels includes pending messages from selected Channels.
	LoadChannels []ChannelDef
	// LoadChannelMaps includes every current ChannelMap instance.
	LoadChannelMaps []ChannelDef
	// LoadChannelMapInstances includes exact ChannelMap instance messages.
	LoadChannelMapInstances []ChannelMapLoad
}

// WorkerTarget identifies the application WorkerService endpoint Dex should call.
type WorkerTarget struct {
	// Address is the advertised plaintext gRPC target, normally host:port.
	Address string
	// Headless bypasses service discovery and requires a direct host:port target.
	Headless bool
}

// ActiveStepSearchMode controls which active Step types Dex indexes for Flow search.
type ActiveStepSearchMode uint8

const (
	// SearchActiveStepsDefault uses the server default indexing policy.
	SearchActiveStepsDefault ActiveStepSearchMode = iota
	// SearchAllActiveSteps indexes every active Step, including execute-only Steps.
	SearchAllActiveSteps
	// SearchActiveStepsWithWaitFor indexes a Step only after WaitFor runs.
	SearchActiveStepsWithWaitFor
	// DisableActiveStepSearch disables active-Step indexing for the Flow.
	DisableActiveStepSearch
)

// FlowConfig overrides mutable runtime behavior for one Flow.
// Nil fields preserve server defaults.
type FlowConfig struct {
	// ActiveStepSearchMode controls active Step indexing.
	ActiveStepSearchMode *ActiveStepSearchMode
	// ContinueAsNewThreshold is the event-count threshold for Continue-As-New.
	ContinueAsNewThreshold *int32
	// ContinueAsNewPageSizeBytes caps history bytes carried into Continue-As-New.
	ContinueAsNewPageSizeBytes *int32
	// StepDurability sets the default persistence durability for Step writes.
	StepDurability *StepDurability
	// WorkerTarget routes future Step and RPC invocations.
	WorkerTarget *WorkerTarget
	// AttributeStoreNames selects Server-configured Attribute Stores for opted-in writes.
	// A nil slice preserves the current targets; an empty non-nil slice disables future projections.
	AttributeStoreNames []string
}

// StartFlowOptions configures a new Flow execution.
// Nil and zero fields preserve server defaults; RequestID is generated when omitted.
type StartFlowOptions struct {
	// Timeout sets Dex's durable soft timeout; nil or zero disables it.
	Timeout *time.Duration
	// TimeoutPolicy controls what Dex does when Timeout expires.
	// The default uses Handler when the Flow implements FlowTimeoutHandler, otherwise Fail.
	TimeoutPolicy FlowTimeoutPolicy
	// TimeoutHandlerOptions configures execution when TimeoutPolicy resolves to TimeoutHandler.
	// It requires a positive Timeout and is invalid with Fail or Cancel.
	TimeoutHandlerOptions *FlowTimeoutHandlerOptions
	// IDReusePolicy controls reuse of an existing Flow ID.
	IDReusePolicy IDReusePolicy
	// StartDelay postpones the first Step after start acceptance.
	StartDelay *time.Duration
	// RetryPolicy configures whole-Flow retries after terminal failures.
	RetryPolicy *FlowRetryPolicy
	// Attributes supplies encoded initial Attribute and Attribute-map values.
	Attributes []InitialAttributeDef
	// ConfigOverride replaces registered Flow configuration for this execution.
	ConfigOverride *FlowConfig
	// AlreadyStarted controls whether an existing Flow is returned as success.
	AlreadyStarted *AlreadyStartedOptions
	// RequestID defaults to a UUID; set a stable business identifier for cross-call retries.
	RequestID *string
}

// FlowTimeoutPolicy controls how a positive Flow timeout ends or redirects execution.
type FlowTimeoutPolicy uint8

const (
	// TimeoutDefault selects Handler for a FlowTimeoutHandler and Fail otherwise.
	TimeoutDefault FlowTimeoutPolicy = iota
	// TimeoutFail fails the Flow with FlowErrorTimeout and permits Flow retries.
	TimeoutFail
	// TimeoutCancel cancels the Flow without retrying it.
	TimeoutCancel
	// TimeoutHandler invokes FlowTimeoutHandler.HandleTimeout as one retryable logical execution.
	TimeoutHandler
)

// SubFlowReusePolicy controls how a generated SubFlow Flow ID resolves an existing execution.
type SubFlowReusePolicy uint8

const (
	// RestartSubFlowIfPreviousExitedAbnormally attaches to running executions, returns completed
	// results, and replaces failed, canceled, timed-out, or terminated executions.
	RestartSubFlowIfPreviousExitedAbnormally SubFlowReusePolicy = iota
	// AttachSubFlow attaches to a running execution or returns its existing terminal result.
	AttachSubFlow
	// AlwaysRestartSubFlow replaces a different existing execution, including a running one.
	AlwaysRestartSubFlow
)

// SubFlowOptions configures one durable SubFlow Condition.
//
// Zero values inherit normal Flow start defaults and use
// RestartSubFlowIfPreviousExitedAbnormally. The server owns the generated Flow ID and request ID;
// callers cannot override either value. Attributes must belong to the target SubFlow.
type SubFlowOptions struct {
	// Timeout limits total SubFlow execution duration.
	Timeout *time.Duration
	// TimeoutPolicy controls what Dex does when Timeout expires.
	// The default uses Handler when the SubFlow implements FlowTimeoutHandler, otherwise Fail.
	TimeoutPolicy FlowTimeoutPolicy
	// TimeoutHandlerOptions configures execution when TimeoutPolicy resolves to TimeoutHandler.
	// It requires a positive Timeout and is invalid with Fail or Cancel.
	TimeoutHandlerOptions *FlowTimeoutHandlerOptions
	// StartDelay postpones the SubFlow starting Step after start acceptance.
	StartDelay *time.Duration
	// RetryPolicy configures whole-Flow retries after terminal failures.
	RetryPolicy *FlowRetryPolicy
	// Attributes supplies encoded initial Attribute and Attribute-map values.
	Attributes []InitialAttributeDef
	// ConfigOverride overrides fields inherited from the parent Flow configuration.
	ConfigOverride *FlowConfig
	// ReusePolicy controls how an execution already using the generated Flow ID is resolved.
	ReusePolicy SubFlowReusePolicy
	// ConditionID assigns the stable ID required by AnyComboOf.
	ConditionID string
}

func defaultSubFlowOptions() SubFlowOptions {
	return SubFlowOptions{ReusePolicy: RestartSubFlowIfPreviousExitedAbnormally}
}

// IDReusePolicy controls whether StartFlow may reuse a Flow ID.
type IDReusePolicy uint8

const (
	// IDReuseDefault uses the server-configured reuse policy.
	IDReuseDefault IDReusePolicy = iota
	// IDReuseAllowIfPreviousFailed permits reuse only after a failed run.
	IDReuseAllowIfPreviousFailed
	// IDReuseAllowIfNotRunning permits reuse when no run is active.
	IDReuseAllowIfNotRunning
	// IDReuseDisallow rejects reuse regardless of previous status.
	IDReuseDisallow
	// IDReuseTerminateIfRunning terminates an active run before replacement.
	IDReuseTerminateIfRunning
)

// FlowRetryPolicy configures whole-Flow retry timing after terminal failure.
type FlowRetryPolicy struct {
	// InitialInterval is the delay before the first Flow retry.
	InitialInterval time.Duration
	// BackoffCoefficient multiplies successive Flow retry intervals.
	BackoffCoefficient float64
	// MaximumInterval caps time between Flow retries.
	MaximumInterval time.Duration
	// MaximumAttempts limits total Flow attempts, including the initial attempt.
	MaximumAttempts int32
}

// AlreadyStartedOptions configures StartFlow behavior when the Flow ID already exists.
type AlreadyStartedOptions struct {
	// IgnoreError returns the existing run ID instead of FlowAlreadyStartedError.
	IgnoreError bool
}

// InvokeOptions configures one RPC invocation.
type InvokeOptions struct {
	// Timeout limits the RPC handler; zero uses its registered or server default.
	Timeout time.Duration
	// LockAttributes are acquired atomically for the RPC invocation.
	LockAttributes []AttributeLock
	// IsTransactional requests transactional reads and writes without requiring Attribute locks.
	// Channel deletions require this option to make a missing message abort all RPC writes.
	IsTransactional bool
	// LoadAttributeMaps includes every current instance of each AttributeMap in the RPC snapshot.
	LoadAttributeMaps []AttributeDef
	// LoadAttributeMapInstances includes exact AttributeMap instances in the RPC snapshot.
	LoadAttributeMapInstances []AttributeMapLoad
	// LoadChannels includes pending messages from the selected Channels in the RPC snapshot.
	LoadChannels []ChannelDef
	// LoadChannelMaps includes every current instance of each ChannelMap in the RPC snapshot.
	LoadChannelMaps []ChannelDef
	// LoadChannelMapInstances includes exact ChannelMap instance messages in the RPC snapshot.
	LoadChannelMapInstances []ChannelMapLoad
}

// WaitForFlowOptions controls Flow-result hydration.
type WaitForFlowOptions struct {
	// NeedsResults asks Dex to include completed Step outputs.
	NeedsResults bool
}

// StopType selects how StopFlow ends an active Flow.
type StopType uint8

const (
	// CancelFlow requests cooperative cancellation.
	CancelFlow StopType = iota + 1
	// TerminateFlow ends the Flow immediately without cleanup.
	TerminateFlow
	// FailFlow marks the Flow failed with Reason.
	FailFlow
)

// StopOptions configures Client.StopFlow.
type StopOptions struct {
	// Type selects cancel, terminate, or fail behavior.
	Type StopType
	// Reason is recorded with the stop operation.
	Reason string
}

// StepExecutionID identifies one execution of a Step type within a Flow run.
type StepExecutionID struct {
	// StepType is the stable registered Step type.
	StepType string
	// ExecutionNumber defaults to one when omitted.
	ExecutionNumber *int32
}

// TimerID selects one Timer Condition within a Step execution.
type TimerID struct {
	// ConditionID targets a Condition configured with WithConditionID.
	ConditionID string
	// Index targets a zero-based Timer Condition position when non-nil.
	Index *int32
}

// TimeTravelType identifies the historical point selector used by Client.TimeTravel.
type TimeTravelType uint8

const (
	// TimeTravelToBeginning resumes before the first history event.
	TimeTravelToBeginning TimeTravelType = iota + 1
	// TimeTravelByHistoryEventTime resumes at the last eligible event by time.
	TimeTravelByHistoryEventTime
	// TimeTravelByStepType resumes before the first execution of a Step type.
	TimeTravelByStepType
	// TimeTravelByStepExecutionID resumes before an exact Step execution.
	TimeTravelByStepExecutionID
)

// TimeTravelStepMethod selects the Step method used as a time travel boundary.
type TimeTravelStepMethod uint8

const (
	// TimeTravelStepWaitFor resumes before the selected Step execution's WaitFor method.
	TimeTravelStepWaitFor TimeTravelStepMethod = iota + 1
	// TimeTravelStepExecute resumes before the selected Step execution's Execute method.
	TimeTravelStepExecute
)

// TimeTravelOptions configures Client.TimeTravel with exactly one historical point.
type TimeTravelOptions struct {
	// Type selects which time travel point field Dex reads.
	Type TimeTravelType
	// Reason is recorded with the time travel operation; empty leaves it unspecified.
	Reason string
	// HistoryEventTime supplies the time for TimeTravelByHistoryEventTime.
	HistoryEventTime time.Time
	// StepType supplies the stable Step type for TimeTravelByStepType.
	StepType string
	// StepExecutionID supplies the exact execution for TimeTravelByStepExecutionID.
	StepExecutionID string
	// StepMethod selects WaitFor or Execute when Type is TimeTravelByStepExecutionID.
	StepMethod TimeTravelStepMethod
	// SkipWritesReapply prevents replay of later RPCs, Channel publications, and Attribute writes; false reapplies them.
	SkipWritesReapply bool
}

File: sdk-typescript/src/options.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import type { Attribute, AttributeMap } from "./persistence.js";
import type { RetryPolicy, StepDurability } from "./step.js";
import { requireName } from "./validation.js";

/** Configures FlowService connectivity and default Flow routing. */
export interface ClientOptions {
  /** Plaintext Dex gRPC target; defaults to `localhost:8801`. */
  readonly serverAddress?: string;
  /** Worker endpoint advertised by `startFlow` unless a Flow overrides it. */
  readonly workerTarget?: WorkerTarget;
}

/** Controls which active Steps are included in Flow search indexing. */
export const ActiveStepSearchMode = Object.freeze({
  /** Uses the Dex server's current default policy. */
  DEFAULT: "default",
  /** Indexes every active Step. */
  ALL: "all",
} as const);

/** Represents a value from {@link ActiveStepSearchMode}. */
export type ActiveStepSearchMode =
  (typeof ActiveStepSearchMode)[keyof typeof ActiveStepSearchMode];

/** Controls whether `startFlow` may reuse a previously used Flow ID. */
export const IdReusePolicy = Object.freeze({
  /** Uses the Dex server's default reuse policy. */
  DEFAULT: "default",
  /** Reuses an ID only after an unsuccessful closed run. */
  ALLOW_IF_PREVIOUS_FAILED: "allowIfPreviousFailed",
  /** Reuses an ID after any closed run. */
  ALLOW_IF_NOT_RUNNING: "allowIfNotRunning",
  /** Terminates an active run before starting the replacement. */
  ALLOW_TERMINATE_IF_RUNNING: "allowTerminateIfRunning",
  /** Rejects every previously used Flow ID. */
  DISALLOW: "disallow",
} as const);

/** Represents a value from {@link IdReusePolicy}. */
export type IdReusePolicy = (typeof IdReusePolicy)[keyof typeof IdReusePolicy];

/** Controls how Dex responds when a positive soft Flow timeout expires. */
export const FlowTimeoutPolicy = Object.freeze({
  /** Invokes `Flow.handleTimeout` when present, and fails otherwise. */
  DEFAULT: "default",
  /** Fails with `FlowErrorType.FLOW_TIMEOUT` and permits Flow retries. */
  FAIL: "fail",
  /** Cancels without retrying the Flow. */
  CANCEL: "cancel",
  /** Invokes `Flow.handleTimeout` once after the durable timer fires or is skipped. */
  HANDLER: "handler",
} as const);

/** Represents a value from {@link FlowTimeoutPolicy}. */
export type FlowTimeoutPolicy =
  (typeof FlowTimeoutPolicy)[keyof typeof FlowTimeoutPolicy];

/** Overrides mutable server behavior for one Flow execution. */
export interface FlowConfig {
  /** Optional active-Step visibility indexing policy. */
  readonly activeStepSearchMode?: ActiveStepSearchMode;
  /**
   * Selects the server-configured Attribute Store receiving opted-in Attribute writes.
   *
   * Omit this property to preserve the existing Flow setting. An explicit empty string is sent with
   * presence and disables later synchronization. Projection is asynchronous and never rolls back Flow
   * Attribute writes when the external store update fails.
   */
  readonly attributeStoreName?: string;
  /** Positive history-event threshold requesting continue-as-new. */
  readonly continueAsNewThreshold?: number;
  /** Positive history page-size budget in bytes for continue-as-new. */
  readonly continueAsNewPageSizeBytes?: number;
  /** Default durability for later Step handler results. */
  readonly stepDurability?: StepDurability;
  /** Worker endpoint used for later handler calls. */
  readonly workerTarget?: WorkerTarget;
}

/**
 * Describes one Attribute value written atomically when a Flow starts.
 * @typeParam T - Attribute value type.
 */
export interface InitialAttribute<T> {
  /** Registered singleton Attribute or AttributeMap definition. */
  readonly attribute: Attribute<T> | AttributeMap<T>;
  /** Required non-empty map key; omitted for a singleton Attribute. */
  readonly instance?: string;
  /** Typed initial value. */
  readonly value: T;
}

/** Creates type-safe initial Attribute writes for {@link StartFlowOptions}. */
export const InitialAttribute = Object.freeze({
  /**
   * Creates a singleton Attribute initialization.
   * @typeParam T - Attribute value type.
   * @param attribute - Registered singleton Attribute.
   * @param value - Typed initial value.
   * @returns The Attribute initialization value.
   */
  of<T>(attribute: Attribute<T>, value: T): InitialAttribute<T> {
    return { attribute, value };
  },
  /**
   * Creates an AttributeMap instance initialization.
   * @typeParam T - Attribute value type.
   * @param attribute - Registered AttributeMap.
   * @param instance - Non-empty logical map key.
   * @param value - Typed initial value.
   * @returns The AttributeMap initialization value.
   */
  mapValue<T>(
    attribute: AttributeMap<T>,
    instance: string,
    value: T,
  ): InitialAttribute<T> {
    requireName(instance);
    return { attribute, instance, value };
  },
});

/**
 * Configures creation of a new Flow execution. All durations use milliseconds.
 *
 * @example
 * ```ts
 * const status = new Attribute("status", stringCodec);
 * const options: StartFlowOptions = {
 *   timeoutMs: 30 * 60_000,
 *   idReusePolicy: IdReusePolicy.ALLOW_IF_NOT_RUNNING,
 *   attributes: [InitialAttribute.of(status, "queued")],
 *   configOverride: { activeStepSearchMode: ActiveStepSearchMode.ALL },
 * };
 * ```
 */
export interface StartFlowOptions {
  /** Dex durable soft timeout in milliseconds; omitted or zero disables it. */
  readonly timeoutMs?: number;
  /** Action taken when a positive timeout expires; defaults from the Flow's hook. */
  readonly timeoutPolicy?: FlowTimeoutPolicy;
  /** Delay before the starting Step becomes eligible, in milliseconds. */
  readonly startDelayMs?: number;
  /** Flow ID reuse policy; uses `DEFAULT` when omitted. */
  readonly idReusePolicy?: IdReusePolicy;
  /** Optional Flow-level retry policy. */
  readonly retryPolicy?: RetryPolicy;
  /** Initial Attribute writes applied atomically with Flow creation. */
  readonly attributes?: readonly InitialAttribute<any>[];
  /** Flow configuration applied over registered defaults. */
  readonly configOverride?: FlowConfig;
  /** Returns the existing run instead of raising an already-started error. */
  readonly ignoreAlreadyStarted?: boolean;
  /** Non-empty idempotency key; generates a UUID when omitted. */
  readonly requestId?: string;
}

/**
 * Describes the lifecycle state of one Flow run.
 * `serverSideTimeoutInternalOnly` is reserved for backend hard-timeout reporting.
 */
export type FlowStatus =
  | "running"
  | "completed"
  | "failed"
  | "cancelled"
  | "terminated"
  | "serverSideTimeoutInternalOnly"
  | "continuedAsNew";

/** Summarizes the current or latest run for one Flow ID. */
export interface FlowInfo {
  /** Stable application-assigned Flow ID. */
  readonly flowId: string;
  /** Server-assigned run ID. */
  readonly runId: string;
  /** Registered Flow type. */
  readonly flowType: string;
  /** Current lifecycle state. */
  readonly status: FlowStatus;
  /** UTC run start timestamp. */
  readonly startedAt: Date;
}

/** Represents one indexed Flow run returned by `searchFlows`. */
export interface SearchFlowEntry {
  /** Stable application-assigned Flow ID. */
  readonly flowId: string;
  /** Server-assigned run ID for this entry. */
  readonly runId: string;
  /** Registered Flow type. */
  readonly flowType: string;
  /** Indexed lifecycle state. */
  readonly status: FlowStatus;
  /** UTC start timestamp. */
  readonly startedAt: Date;
  /** UTC close timestamp, or `undefined` for open or unknown runs. */
  readonly closedAt: Date | undefined;
  /** Hydrated values keyed by physical search-index name. */
  readonly indexedAttributes: ReadonlyMap<string, unknown>;
}

/** Contains one server-ordered page of Flow search results. */
export interface SearchFlowsPage {
  /** Result entries in server-defined query order. */
  readonly flows: readonly SearchFlowEntry[];
  /** Opaque next-page token, or an empty string on the final page. */
  readonly nextPageToken: string;
}

/** Identifies one numbered execution of a registered Step type. */
export interface StepExecutionId {
  /** Registered Step type. */
  readonly stepType: string;
  /** Positive execution number; defaults to the first execution. */
  readonly number?: number;
}

/** Creates normalized {@link StepExecutionId} values. */
export const StepExecutionId = Object.freeze({
  /**
   * Creates a Step execution identifier.
   * @param stepType - Non-empty registered Step type.
   * @param number - Positive execution number; defaults to one.
   * @returns The normalized identifier.
   */
  of(stepType: string, number?: number): StepExecutionId {
    return { stepType, number: number ?? 1 };
  },
});

/** Selects one Timer condition within a Step execution. */
export interface TimerId {
  /** Stable condition ID, mutually exclusive with `conditionIndex`. */
  readonly conditionId?: string;
  /** Zero-based condition index, mutually exclusive with `conditionId`. */
  readonly conditionIndex?: number;
}

/** Creates Timer selectors by stable ID or Wait-tree position. */
export const TimerId = Object.freeze({
  /**
   * Selects a Timer by stable condition ID.
   * @param conditionId - Non-empty ID assigned by `Timer.byDuration`.
   * @returns An ID-based Timer selector.
   */
  byConditionId(conditionId: string): TimerId {
    requireName(conditionId);
    return { conditionId };
  },
  /**
   * Selects a Timer by zero-based Wait-tree position.
   * @param conditionIndex - Non-negative flattened Timer index.
   * @returns An index-based Timer selector.
   */
  byConditionIndex(conditionIndex: number): TimerId {
    return { conditionIndex };
  },
});

/** Selects the historical point from which time travel creates a new run. */
export const TimeTravelType = Object.freeze({
  /** Restarts from the beginning of the Flow. */
  BEGINNING: "beginning",
  /** Resumes at the first event at or after a timestamp. */
  HISTORY_EVENT_TIME: "historyEventTime",
  /** Resumes at the latest execution of a Step type. */
  STEP_TYPE: "stepType",
  /** Resumes at one exact Step execution ID. */
  STEP_EXECUTION_ID: "stepExecutionId",
} as const);

/** Represents a value from {@link TimeTravelType}. */
export type TimeTravelType = (typeof TimeTravelType)[keyof typeof TimeTravelType];

/** Selects the Step method used as a Step execution time travel boundary. */
export const TimeTravelStepMethod = Object.freeze({
  /** Reruns WaitFor and everything after it. */
  WAIT_FOR: "waitFor",
  /** Keeps the WaitFor result and reruns Execute and everything after it. */
  EXECUTE: "execute",
} as const);

/** Represents a value from {@link TimeTravelStepMethod}. */
export type TimeTravelStepMethod = (typeof TimeTravelStepMethod)[keyof typeof TimeTravelStepMethod];

/** Configures creation of a new run from existing Flow history. */
export interface TimeTravelOptions {
  /** Time travel point selector kind. */
  readonly type: TimeTravelType;
  /** Timestamp used with `HISTORY_EVENT_TIME`. */
  readonly historyEventTime?: Date;
  /** Registered Step type used with `STEP_TYPE`. */
  readonly stepType?: string;
  /** Exact execution ID used with `STEP_EXECUTION_ID`. */
  readonly stepExecutionId?: string;
  /** WaitFor or Execute boundary required with `STEP_EXECUTION_ID`. */
  readonly stepMethod?: TimeTravelStepMethod;
  /** Optional operator-readable time travel reason. */
  readonly reason?: string;
  /** Prevents reapplication of RPCs, Channel publications, and Attribute writes after the selected point. */
  readonly skipWritesReapply?: boolean;
}

/** Selects how an active Flow should close. */
export const StopType = Object.freeze({
  /** Requests cooperative cancellation. */
  CANCEL: "cancel",
  /** Forces immediate termination. */
  TERMINATE: "terminate",
  /** Closes the Flow as failed. */
  FAIL: "fail",
} as const);

/** Represents a value from {@link StopType}. */
export type StopType = (typeof StopType)[keyof typeof StopType];

/** Configures a `stopFlow` request. */
export interface StopFlowOptions {
  /** Stop behavior; defaults to cooperative cancellation. */
  readonly type?: StopType;
  /** Optional operator-readable reason recorded by Dex. */
  readonly reason?: string;
}

/** Configures Worker networking and startup synchronization. */
export interface WorkerOptions {
  /** Local plaintext WorkerService listener; defaults to `:8803`. */
  readonly bindAddress?: string;
  /** Endpoint advertised to Dex; derived from `bindAddress` when omitted. */
  readonly workerTarget?: WorkerTarget;
  /** Dex FlowService target; defaults to `localhost:8801`. */
  readonly serverAddress?: string;
  /** Startup Indexed Attribute synchronization deadline in milliseconds. Defaults to 120000. */
  readonly attributeIndexSyncTimeoutMs?: number;
}

/** Identifies the application Worker endpoint advertised to Dex. */
export interface WorkerTarget {
  /** Plaintext gRPC target; headless targets must use `host:port`. */
  readonly address: string;
  /** Whether Dex connects directly without service discovery. */
  readonly headless?: boolean;
}
